File: google/main/src/utils/ConsoleApp.java
package main.src.utils;

import java.util.*;


public abstract class ConsoleApp {

    protected String prompt;
    protected Scanner scanner;

    private String command;
    private String arguments;
    private int firstSpaceIndex;
    

    public abstract void intro();

    public abstract void help();

    protected abstract boolean handleCommand(String command, String arguments);

    public ConsoleApp(){
        this.intro();
        this.help();
        this.scanner = new Scanner(System.in);
        this.prompt = "";
    }

    private void decomposeInput(String input){
        this.firstSpaceIndex = input.indexOf(' ');
        if(firstSpaceIndex < 0){
            this.firstSpaceIndex = input.length();
            this.command = input;
            this.arguments = "";
        }
        else{
            this.command = input.substring(0, this.firstSpaceIndex).trim();
            this.arguments = input.substring(this.firstSpaceIndex + 1).trim();
        }
    }

    public void run() {
        String input;
        do {
            System.out.print(this.prompt);
            input = scanner.nextLine().trim();
            this.decomposeInput(input);
        } while(handleCommand(this.command, this.arguments));
    }
}

File: GoogleSharpApp/SearchEngine/ConsoleApp.cs
using System;

namespace SearchEngine {
    public abstract class ConsoleApp {

        protected string prompt;
        protected string command;
        protected string arguments;
        protected int firstSpaceIndex;

        public abstract void Intro();

        public abstract void Help();

        public abstract bool HandleCommand();

        public ConsoleApp() {
            Intro();
            Help();
        }

        private void DecomposeInput(string input) {
            firstSpaceIndex = input.IndexOf(' ');
            if (firstSpaceIndex < 0) {
                firstSpaceIndex = input.Length;
                command = input;
                arguments = "";
            } else {
                command = input.Substring(0, firstSpaceIndex).Trim();
                arguments = input.Substring(firstSpaceIndex + 1).Trim();
            }
        }

        public void Run() {
            string input;
            do {
                Console.Write(prompt);
                input = Console.ReadLine().Trim();
                DecomposeInput(input);
            } while (HandleCommand());
        }
    }
}
